www/src/content/blog/2025-05-21-rate-limiting-hono-apps-intro.mdx

```diff
@@ -0,0 +1,291 @@
+
+In the next article, I’ll introduce more flexible and accurate options. Their increased efficacy comes at the cost of logical complexity and performance though, so for many use-cases the Fixed Window algorithm’s strengths outweigh its weaknesses.
+
+### Calibrating rate limits
+
+The appropriate configuration for a rate limiter depends largely on the problem it’s meant to solve. Each use-case has distinct priorities and constraints, and effective rate limits are the product of balancing one against the other.
+
+Limiters enforcing tiered API access balance resource costs against subscription prices, while those protecting auth routes prioritize security while allowing for some user error. Likewise, limiters preventing resource exhaustion should reflect average consumption relative to capacity.
+
+When first adding a rate limiter, it’s best to choose a more conservative limit in order to ensure that all users have fair access and that service isn’t disrupted. This initial configuration can then be tuned over time to better reflect actual request patterns.
+
+Sophisticated rate limiters may use dynamic limits to account for fluctuations throughout the day, or to handle traffic spikes related to planned events like a promotion or content release. In most cases though, using server metrics to regularly tune static rate limits is good enough.
+
+## Uniquely identifying clients
+
+Both the standard and Cloudflare-specific middleware require a `keyGenerator`, which takes Hono `Context` as an argument and must return a string key. This key is used to distinguish between records in the limiter store, so for most use-cases it should be unique to each client.
+
+While many rate limiting examples use the client IP for convenience, this isn’t recommended for production apps. IP addresses aren’t guaranteed to be unique, and bad actors can easily spoof IPs or obscure them with a VPN. 
+
+A user or account ID is the best option for most use-cases, since their uniqueness makes it easier to apply rate limits fairly and accurately. This is essential for premium APIs and other subscription services, which promise customers a defined level of access for each payment tier.
+
+### When to rate limit by IP
+
+Rate limiters protecting pre-login or unprotected routes are a notable exception. Uniquely identifying clients is more difficult in these cases, but also less relevant to the limiters’ role, which is focused on security and capacity management more-so than regulating individual client usage.
+
+In auth workflows, rate limiters make attempts to gain unauthorized access to the system logistically challenging or prohibitively exepensive. Since users only need to submit their credentials once, any IPs that exceed a reasonable rate limit are suspicious, and may even be blocked after repeated violations.
```

Comment:
```suggestion
In auth workflows, rate limiters make attempts to gain unauthorized access to the system logistically challenging or prohibitively expensive. Since users only need to submit their credentials once, any IPs that exceed a reasonable rate limit are suspicious, and may even be blocked after repeated violations.
```